### ISSUE TITLE
False alarm: TRIO107 for abstractmethods

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/flake8-trio#40]
```python
from abc import ABC, abstractmethod


class Handle(ABC):
    @abstractmethod
    async def ajoin(self):
        ...
```
```
$ flake8 example.py
example.py:6:5: TRIO107: exit from async function with no guaranteed checkpoint or exception since function definition on line 6
```

Note that we'll want to handle abstractclassmethod and abstractproperty too.  In addition to the explicit decorator-based handling, I'd skip 107 for functions where `...` is the complete body.

**Comment by jakkdl** [comment]
how about functions with `pass` as the complete body?

**Comment by jakkdl** [comment]
and if checking for `...`, is there any reason to check for the decorators explicitly?

**Comment by Zac-HD** [comment]
Yeah, let's treat pass the same as ellipses. (I dislike the style, but whatever)

And we still need to handle the decorators because it's fairly common to put a `raise NotImplementedError`... actually we could handle that (and optionally with a message) the same way too. Still worth handling via decorators in case of eg `msg = "long string" ; raise NotImplementedError(msg)` or something though, which doesn't fit the one-statement heuristic.

**Comment by jakkdl** [comment]
if the body is only a `raise` it's fine, since raises are treated as checkpoints.
`tests/trio107.py:248`
```python
async def foo_raise_1():  # safe
    raise ValueError()
```
and any extra statements with strings or variables or stuff won't bother that.

In theory an abstract class can contain plenty code that gets executed though, due to subclasses calling them through `super()` and the like:

```python
class Parent:
  @abstractmethod
  async def foo(self):
      for i in range(10):
          yield i

class Child(Parent):
  async def foo(self):
    return tuple(super().foo())
```
(invalid python code, but you get what I mean)

So unless you have more cases it feels like the right thing to do is ignore the decorators and just check for ellipsis or pass?

**Comment by Zac-HD** [comment]
> So unless you have more cases it feels like the right thing to do is ignore the decorators and just check for ellipsis or pass?

👍

**Comment by jakkdl** [comment]
Changed it to be `any statement in body that's not a string, pass or ellipsis` - and when tinkering with test cases I realized that pyright has the same check for overloaded functions `error: <> is marked as overloaded, but it includes an implementation` and when digging into it turns out they have the exact same logic ^^
https://github.com/microsoft/pyright/blob/6fa755c025cb5b18838d8bfdd85761184afa45f1/packages/pyright-internal/src/analyzer/parseTreeUtils.ts#L1999

**Comment by Zac-HD** [comment]
Niiice. Leave this comment in the code too!
